scylla-cdc/src/log_reader.rs

```diff
@@ -141,10 +141,25 @@ impl CDCReaderWorker {
                             })
                         })
                         .collect();
-                } else if had_first_generation {
-                    // FIXME: There may be another generation coming in the future with timestamp < end_timestamp
-                    // that could have been missed because of earlier fetching failures.
-                    // More on this here https://github.com/piodul/scylla-cdc-rust/pull/10#discussion_r826865162
+                } else if let Some(last_generation) = last_fetched_generation.take() {
+                    if let Ok(Some(generation)) =
+                        fetcher.fetch_next_generation(&last_generation).await
+                    {
+                        if generation.timestamp <= self.end_timestamp {
+                            // Fetched next generation with lower or equal timestamp than end_timestamp
+                            next_generation = Some(generation.clone());
+                            self.set_upper_timestamp(generation.timestamp).await;
+                            continue;
+                        } else {
+                            // Fetched next generation with bigger timestamp than end_timestamp
+                            return Ok(());
+                        }
+                    } else {
+                        // Last fetched generation's next generation is not available
+                        return Ok(());
```

Comment:
We should at least print a warning here, describing what happened and warning about potentially unread rows.